source/mir/random/ndvariable.d

```diff
@@ -279,6 +280,114 @@ nothrow @safe version(mir_random_test) unittest
     assert(fabs(x[0] + x[1] + x[2] - 1) < 1e-10);
 }
 
+/++
+Multinomial distribution.
++/
+struct MultinomialVariable(U:size_t, T)
+    if (isFloatingPoint!T)
+{
+    import mir.random.variable : binomialVar;
+
+    ///
+    enum isNdRandomVariable = true;
+    ///
+    alias Element = uint;
+
+    ///
+    const(T)[] probs;
+    size_t N;
+
+    /++
+    Params:
+        probs = probabilities of the multinomial distribution
+        N = Number of rolls
+    Constraints: `sum(probs[i]) <= 1`
+    +/
+    this()(size_t N, const(T)[] probs)
+    {
+        this.N = N;
+        this.probs = probs;
+    }
+
+    ///
+    pragma(inline, false)
+    void opCall(G)(scope ref G gen, scope U[] result)
+        if (isSaturatedRandomEngine!G)
+    {
+        double norm = 0.0;
+        double sum_p = 0.0;
+
+        uint sum_n = 0;
+
+
+        /// Makes sure probabilities add up to one, by calculating a normalization term
+        foreach(k, p; this.probs)
```

Comment:
Lets move this loop into the constructor and check with `assert(fabs(norm - 1) <= double.min_normal * probs.length * 2)`

Reply:
We currently don't have a constructor, at least not an explicit one: `this(){...}`
Should we add one?

Reply:
This assert is not correct. `norm` is the sum of values passed in the probs array. If they don't add to one, the code in the main loop, `p / (norm - sum_p)` will make sure `p`s add to one.  The test in the second unittest is easy to calculate by hand to verify this.

Reply:
But I have moved the loop into the constructor. It cleans up the main calling function by removing a computation that only need to be done on initialization.

Reply:
@fccoelho I was initially sympathetic to @9il's point here. If norm is different from 1, then there must be some reason for that. I would rather the user be alerted to the issue and solve the problem on their own. However, I also noticed that the code itself does not have any restriction that the p's are less than 1. I don't see anything that would prevent passing something like p = [10, 30, 60] to Multinomial. I'm not sure if this is intended or not.

Reply:
Well, it seems that some protection against probs that don't add up to one is necessary that's what the norm term is used for. One motivation I can think of for being able to parameterize with arbitrary numbers (which don't add up to one), would be to simplify creating an equivalent multinomial generator from a sample generated by another one, of which you don't necessarily know the probs.

Reply:
From the API point of view the probs should be already normalized. Otherwise it is weights. Mir is positioned as a library with low level abilities and maximal control for a user to create his own wrappers on top of that. Probs are preferred.

Reply:
We can simply assert probs sum up to 1, but then we expose the call to a runtime error if probs are not carefully specified. The way it is implemented right now, it will never throw such an error. So it would be safer without any extra computational cost, since for the assert we would have to sum the probs anyway. Besides, there would be no impact to users who wish to pass exact probabilities.

Reply:
Just place the loop and assert itself into `version(assert)` block
```
version(assert)
{
 /// foreach loop here
}
```

Reply:
Ok, but how do I specify in the unittests which version should be tested?
By the way, is this the way it should be done?
```D
this()(size_t N, const(T)[] probs)
    {
        this.N = N;
        this.probs = probs;
         /// Makes sure probabilities add up to one, by calculating a normalization factor
        version(assert)
        {
            foreach(k, p; this.probs)
        {
            this.norm += p;
        }
            assert(fabs(norm - 1) <= double.min_normal * probs.length * 2);
        }
        else
        {
            foreach (k, p; this.probs)
            {
                this.norm += p;
            }
        }
        //assert(fabs(this.norm - 1) <= double.min_normal * this.probs.length * 2);
    }
```

Reply:
```d
    this()(size_t N, const(T)[] probs)
    {
        this.N = N;
        this.probs = probs;
         /// Makes sure probabilities add up to one, by calculating a normalization factor
        version(assert)
        {
            T norm = 0;
            foreach(k, p; this.probs)
            {
                norm += p;
            }
            assert(fabs(norm - 1) <= double.min_normal * probs.length * 2);
        }
    }
```
`version(assert)` is always set when asserts are set.

Reply:
There's another issue: a conservative assumption about the rounding error. See this example:

https://run.dlang.io/is/cVUDcv

The definition of the probs seems fine, but the rounding error is much greater than what is allowed by the assert. The rounding error, `fabs(norm-1)`,  grows much faster than `double.min_normal * probs.length * 2`. The plot below shows the rounding error from probs lengths from 3 to 200: in orange and the function above in green.

![image](https://user-images.githubusercontent.com/140123/67486379-4f497780-f642-11e9-80db-6490410d5ac4.png)

I think this is a strong reason for not adding this assert.

Reply:
> Thank you for the work on this MR!

You're welcome, it's been a good opportunity for me to learn a bit of D.

I have replaced `min_normal` by `epsilon` and it's now guaranteed to cover the rounding errors.